File: implementations/typescript/elp_omega.test.ts
import { EntangledLogicOmega, Reality } from './elp_omega';

describe('ELP-Omega TypeScript Implementation', () => {
    const secret = 'vortex-secret';
    const elp = new EntangledLogicOmega(secret);

    test('should validate Zeckendorf non-adjacency', () => {
        expect(elp.isValidZeckendorf(0b1001)).toBe(true);
        expect(elp.isValidZeckendorf(0b0011)).toBe(false);
    });

    test('should return PRIME for valid request', () => {
        const ts = Date.now();
        const req = {
            mask: 1,
            context: 'test',
            timestamp: ts,
            path: '/api',
            nonce: 'n1'
        };
        const seal = elp.computeSeal(req);
        const result = elp.processRequest({ ...req, seal }, 'SECRET', 'fp1');
        expect(result.reality).toBe(Reality.PRIME);
    });

    test('should trigger SHADOW on nonce replay', () => {
        const ts = Date.now();
        const req = { mask: 1, context: 'c', timestamp: ts, path: '/p', nonce: 'r1' };
        const seal = elp.computeSeal(req);
        
        elp.processRequest({ ...req, seal }, 'D', 'fp');
        const result = elp.processRequest({ ...req, seal }, 'D', 'fp');
        
        expect(result.reality).toBe(Reality.SHADOW);
    });
});

File: implementations/go/elp_omega_test.go
package elpomega

import (
	"encoding/base64"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestZeckendorfValidation(t *testing.T) {
	elp := NewELP([]byte("test-secret"))

	// 9 = 1001 (Ok)
	assert.True(t, elp.isValidZeckendorfMask(9), "Read + Admin deve ser válido")
	// 3 = 0011 (Fail - Adjacente)
	assert.False(t, elp.isValidZeckendorfMask(3), "Read + Write deve ser inválido")
}

func TestPrimeRealityFlow(t *testing.T) {
	secret := []byte("vortex-secret")
	elp := NewELP(secret)

	req := SecureRequest{
		ZeckendorfMask: 9,
		Context:        "ctx",
		Timestamp:      time.Now().UnixMilli(),
		Path:           "/api",
		Nonce:          "unique-123",
	}
	req.Seal = elp.ComputeSeal(req)

	result, reality := elp.ProcessRequest(req, "DATA", "fp")
	
	assert.Equal(t, RealityPrime, reality)
	assert.Contains(t, result, "PRIME_REALITY")
}

func TestShadowRealityOnReplay(t *testing.T) {
	elp := NewELP([]byte("secret"))
	req := SecureRequest{
		ZeckendorfMask: 9,
		Context:        "ctx",
		Timestamp:      time.Now().UnixMilli(),
		Path:           "/api",
		Nonce:          "replay-nonce",
	}
	req.Seal = elp.ComputeSeal(req)

	// 1ª vez: OK
	elp.ProcessRequest(req, "DATA", "fp")
	
	// 2ª vez: Replay -> Shadow
	res2, r2 := elp.ProcessRequest(req, "DATA", "fp")
	assert.Equal(t, RealityShadow, r2)
	assert.Contains(t, res2, "SHADOW_VAULT_ID")
}

func TestShadowFormat(t *testing.T) {
	elp := NewELP([]byte("s"))
	shadow := elp.generateShadow("d", "c", "p")
	
	// Verifica se é URL Safe Base64
	_, err := base64.RawURLEncoding.DecodeString(shadow)
	assert.NoError(t, err)
	assert.False(t, strings.Contains(shadow, "+"))
}